### PR TITLE
DQM Serialization

### PR DESCRIPTION
**Pull request by arcondello**
Also adds intermediate representation in `DQM.to_numpy_vectors` and `.from_numpy_vectors`.

Also uses the API suggested in #599.

### CODE REVIEW

File: tests/test_discrete_quadratic_model.py [hunk 1]
```diff
@@ -13,12 +13,44 @@
...
+import tempfile
 import unittest
 
 import dimod
 import numpy as np
 
 
+# will want to migrate this to generators at some point
+def gnp_random_dqm(num_variables, num_cases, p, p_case, seed=None):
+    np.random.seed(seed)
```
Comment (randomir) [inline tests/test_discrete_quadratic_model.py:27]: To limit function side-effects, you can use [`numpy.random.default_rng`](https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.default_rng).

File: tests/test_discrete_quadratic_model.py [hunk 2]
```diff
@@ -107,6 +139,96 @@ def test_two_variable_labelled(self):
...
+            new = dimod.DQM.from_file(tf)
+
+        self.assertEqual(new.num_variables(), dqm.num_variables())
+        self.assertEqual(new.num_cases(), dqm.num_cases())
+        self.assertEqual(new.variables, dqm.variables)
+        for v in dqm.variables:
+            self.assertEqual(new.num_cases(v), dqm.num_cases(v))
+            np.testing.assert_array_equal(new.get_linear(v),
+                                          dqm.get_linear(v))
+        self.assertEqual(new.adj, dqm.adj)
```
Comment (randomir) [inline tests/test_discrete_quadratic_model.py:175]: `DQM.__eq__`?

File: dimod/discrete/cydiscrete_quadratic_model.pyx [hunk 3]
```diff
@@ -109,6 +105,142 @@ cdef class cyDiscreteQuadraticModel:
...
+        # if necessary
+        index_dtype = np.result_type(starts, irow, icol, np.uint16)
+
+        starts = np.asarray(starts, dtype=index_dtype)
+        ldata = np.asarray(ldata, dtype=obj.dtype)
+        irow = np.asarray(irow, dtype=index_dtype)
+        icol = np.asarray(icol, dtype=index_dtype)
+        qdata = np.asarray(qdata, dtype=obj.dtype)
+
+        if index_dtype == np.uint16:
```
Comment (randomir) [inline dimod/discrete/cydiscrete_quadratic_model.pyx:227]: No `(u)int8`?

File: dimod/discrete/discrete_quadratic_model.py [hunk 4]
```diff
@@ -383,5 +521,210 @@ def set_quadratic_case(self, u, u_case, v, v_case, bias):
...
+
+        if not index_labeled:
+            file.write(VariablesSection(self.variables).dumps())
+
+        file.seek(0)
+
+        return file
+
+    def to_numpy_vectors(self, return_labels=False):
+        """Convert the DQM to five numpy vectors.
```
Comment (randomir) [inline dimod/discrete/discrete_quadratic_model.py:683]: Just a thought: maybe it's time to start returning `namedtuple`.

**Review by randomir — APPROVED** [review]
LGTM

File: dimod/discrete/cydiscrete_quadratic_model.pyx (hunk 3 above)
Comment (arcondello) [reply dimod/discrete/cydiscrete_quadratic_model.pyx:227]: A few lines above I coerce it to be at least `np.uint16`

File: tests/test_discrete_quadratic_model.py (hunk 1 above)
Comment (arcondello) [reply tests/test_discrete_quadratic_model.py:27]: Totally, I was in a hurry and this one lives only in testing. But good point, might as well to it correctly :smile:

File: tests/test_discrete_quadratic_model.py (hunk 2 above)
Comment (arcondello) [reply tests/test_discrete_quadratic_model.py:175]: The problem comes with handling differently ordered DQMs. Right now `bqm == other_bqm` returns `True` even when they have a different variable order. I would probably want to be consistent with BQM, but I wanted to think about it a bit more first. In either case this (mostly) tests the stronger condition that they are equal and have identical underlying values.

File: dimod/discrete/discrete_quadratic_model.py (hunk 4 above)
Comment (arcondello) [reply dimod/discrete/discrete_quadratic_model.py:683]: Yeah, I was trying to maintain similarity to `BQM.to_numpy_vectors`, but probably practicality trumps similarity. `namedtuple` seems cleaner. Maybe I'll add an option to return a "flat" `namedtuple` from the BQM ones as well as an argument and put up a deprecation notice.

File: dimod/discrete/cydiscrete_quadratic_model.pyx (hunk 3 above)
Comment (randomir) [reply dimod/discrete/cydiscrete_quadratic_model.pyx:227]: Missed that. :+1:

File: dimod/discrete/discrete_quadratic_model.py (hunk 4 above)
Comment (randomir) [reply dimod/discrete/discrete_quadratic_model.py:683]: No need, `namedtuple` is `tuple`'s subclass, so existing code would not break. At least destructuring will work. Doubt anyone checks exact rtype.

Comment (arcondello) [reply dimod/discrete/discrete_quadratic_model.py:683]: I was more referring to the fact that I returned a nested tuple `(case_starts, linear_biases, (irow, icol, qdata))`. In either case, I can just return a nested namedtuple as well, which is what I will do. I think if I was writing it from scratch I would just return a flat one but :shrug:

Comment (randomir) [reply dimod/discrete/discrete_quadratic_model.py:683]: Oh, I see. Well, we'll fix it in Ocean 5. :smile:

**Pull request merged**